src/Features/Core/Portable/CodeStyle/AbstractCodeStyleProvider.cs

```diff
@@ -0,0 +1,141 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.CodeStyle
+{
+    // This file contains the "protected" surface area of the AbstractCodeStyleProvider.
+    // It specifically is all the extensibility surface that a subclass needs to fill in
+    // in order to properly expose a code style analyzer/fixer/refactoring.
+
+    /// <summary>
+    /// This is the core class a code-style feature needs to derive from.  All logic related to the
+    /// feature will then be contained in this class.  This class will take care of many bit of
+    /// common logic that all code style providers would have to care about and can thus do that
+    /// logic in a consistent fashion without all providers having to do the same.  For example,
+    /// this class will check the current value of the code style option. If it is 'refactoring
+    /// only', it will not bother running any of the DiagnosticAnalyzer codepaths, and will only run
+    /// the CodeRefactoringProvider codepaths.
+    /// </summary>
+    internal abstract partial class AbstractCodeStyleProvider<
+        TOptionKind, TCodeStyleProvider>
+        where TCodeStyleProvider : AbstractCodeStyleProvider<TOptionKind, TCodeStyleProvider>, new()
+    {
+        private readonly Option<CodeStyleOption<TOptionKind>> _option;
+        private readonly string _descriptorId;
+        private readonly LocalizableString _title;
+        private readonly LocalizableString _message;
+
+        protected AbstractCodeStyleProvider(
+            Option<CodeStyleOption<TOptionKind>> option,
+            string descriptorId,
+            LocalizableString title,
+            LocalizableString message)
+        {
+            _option = option;
+            _descriptorId = descriptorId;
+            _title = title;
+            _message = message;
+        }
+
+        /// <summary>
+        /// Helper to get the true ReportDiagnostic severity for a given option.  Importantly, this
+        /// handle ReportDiagnostic.Default and will map that back to the appropriate value in that
+        /// case.
+        /// </summary>
+        protected static ReportDiagnostic GetOptionSeverity(CodeStyleOption<TOptionKind> optionValue)
+        {
+            var severity = optionValue.Notification.Severity;
+            return severity == ReportDiagnostic.Default
+                ? severity.WithDefaultSeverity(DiagnosticSeverity.Hidden)
+                : severity;
```

Comment:
@sharwell said this is the right thing to do with ReportDiagnostic is set to Default.

Reply:
gotcha.  I thought from a previous discussion you recommended the above in case that behavior would change.  But i'm happy to simplify if you think it's the right thing to do.